elfloader-tool/src/common.c

```diff
@@ -618,3 +623,41 @@ WEAK void platform_init(void)
 {
     /* nothing by default */
 }
+
+#ifdef CONFIG_ARCH_ARM
+extern void flush_dcache(void);
+extern void invalidate_icache(void);
+#endif
+
+/*
+ * Moves the elfloader to its expected start address in memory.
+ * This is for when the elfloader needs to be in a specific location
+ * to ensure that it doesn't occupy memory that it needs to load the
+ * kernel and user applications into.
+ */
+char * fixup_image_base(void) {
+    char *load_base = _start;
+    char *target_base = (char *) IMAGE_START_ADDR;
+    if (load_base == target_base) {
+        /* already in the right place, just keep booting */
+        return (char *)0;
+    }
+
+    /* Check that the current image location doesn't overlap with the
```

Comment:
we could copy backwards if they overlap, just not sure if that is necessary.

Reply:
Actually the memmove implementation in the elfloader can handle this. I originally thought that it was just the same implementation of memcpy.  I can fix this up to handle overlap apart from the code segment apart so that it's like the original assembly (which I originally thought was not properly handling overlap).

Reply:
Isn't the issue when there is overlapping, there is chance that the code that is currently executing is not overwritten and thus the result might be undefined? So this must be detected and avoided. Same for overwriting the stack/global in bss

Reply:
I've fixed this up to only check for collision up until _archive_start.